e2e/pvc_test.go

```diff
@@ -66,107 +67,71 @@ var _ = Context("pvc_test", func() {
-		Expect(err).NotTo(HaveOccurred())
-
-		wantLabels := make(map[string]map[string]string)
-		for _, pvc := range cluster.Spec.VolumeClaimTemplates {
-			wantLabels[pvc.Name] = pvc.ObjectMeta.Labels
-		}
+		verifyPVCTemplates("pvc", "cluster")
+	})
 
-		wantSizes := make(map[string]*resource.Quantity)
-		for _, pvc := range cluster.Spec.VolumeClaimTemplates {
-			wantSizes[pvc.Name] = pvc.Spec.Resources.Requests.Storage()
-		}
+	It("should pvc resized", func() {
+		verifyPVCSize("pvc", "cluster")
+	})
 
+	It("should pvc template storage size reduce succeed", func() {
+		// 1Gi -> 500Mi
+		kubectlSafe(fillTemplate(pvcTestYAML), "apply", "-f", "-")
 		Eventually(func() error {
-			out, err := kubectl(nil,
-				"get", "sts",
-				"-n", "pvc",
-				"moco-cluster",
-				"-o", "json",
-			)
+			cluster, err := getCluster("pvc", "cluster")
 			if err != nil {
 				return err
 			}
-
-			var sts appsv1.StatefulSet
-			if err := json.Unmarshal(out, &sts); err != nil {
-				return err
-			}
-
-			for _, pvc := range sts.Spec.VolumeClaimTemplates {
-				labels, ok := wantLabels[pvc.Name]
-				if !ok {
-					return fmt.Errorf("pvc %s is not expected", pvc.Name)
-				}
-
-				if !reflect.DeepEqual(pvc.ObjectMeta.Labels, labels) {
-					return fmt.Errorf("pvc %s labels are not expected", pvc.Name)
-				}
-
-				want, ok := wantSizes[pvc.Name]
-				if !ok {
-					return fmt.Errorf("pvc %s is not expected", pvc.Name)
+			for _, cond := range cluster.Status.Conditions {
+				if cond.Type != mocov1beta2.ConditionHealthy {
+					continue
 				}
-
-				if pvc.Spec.Resources.Requests.Storage().Cmp(*want) != 0 {
-					return fmt.Errorf("pvc %s is not expected size: %s", pvc.Name, pvc.Spec.Resources.Requests.Storage())
+				if cond.Status == corev1.ConditionTrue {
+					return nil
 				}
+				return fmt.Errorf("cluster is not healthy: %s", cond.Status)
 			}
-
-			return nil
+			return errors.New("no health condition")
 		}).Should(Succeed())
 	})
 
-	It("should pvc resized", func() {
-		cluster, err := getCluster("pvc", "cluster")
-		Expect(err).NotTo(HaveOccurred())
-
-		wantSizes := make(map[string]*resource.Quantity)
-		for _, pvc := range cluster.Spec.VolumeClaimTemplates {
-			for i := int32(0); i < cluster.Spec.Replicas; i++ {
-				name := fmt.Sprintf("%s-%s-%d", pvc.Name, "moco-cluster", i)
-				wantSizes[name] = pvc.Spec.Resources.Requests.Storage()
-			}
-		}
-
-		Eventually(func() error {
-			out, err := kubectl(nil,
-				"get", "pvc",
-				"-n", "pvc",
-				"-l", "app.kubernetes.io/instance=cluster",
-				"-o", "json",
-			)
-			if err != nil {
-				return err
-			}
+	It("should statefulset re-created", func() {
+		verifyPVCTemplates("pvc", "cluster")
+	})
 
-			var pvcList corev1.PersistentVolumeClaimList
-			if err := json.Unmarshal(out, &pvcList); err != nil {
-				return err
-			}
-			if len(pvcList.Items) < 1 {
-				return errors.New("not found pvcs")
-			}
+	It("should volume size reduce succeed", func() {
```

Comment:
Can you confirm that the data is not lost after this test?

Reply:
Thanks, I added test to check for data loss.
https://github.com/cybozu-go/moco/pull/552/commits/57652b63c8237fa2a2fe4eaf2a8086a73507d0b1